contracts/Tokens/VTokens/VToken.sol

```diff
@@ -351,6 +344,168 @@ abstract contract VToken is VTokenInterface, Exponential, TokenErrorReporter {
+     *      - FlashLoans must be enabled for the asset.
+     *      - The `receiver` contract must repay the loan with the appropriate fee.
+     * @param initiator The address that initiated the flash loan.
+     * @param receiver The address of the contract that will receive the flashLoan and execute the operation.
+     * @param amount The amount of asset to be loaned.
+     * @param param Additional encoded parameters passed with the flash loan.
+     * @custom:error FlashLoanNotEnabled is thrown if flash loans are disabled for the asset.
+     * @custom:error SenderNotAuthorized is thrown if the initiator is not authorized to execute flash loan.
+     * @custom:error ExecuteFlashLoanFailed is thrown if the receiver contract fails to execute the operation.
+     * @custom:error InsufficientRepayment is thrown if the repayment (amount + fee) is insufficient after the operation.
+     * @custom:event Emits FlashLoanExecuted event on success
+     */
+    function executeFlashLoan(
+        address initiator,
+        address payable receiver,
+        uint256 amount,
+        bytes calldata param
+    ) external nonReentrant returns (uint256) {
+        if (!isFlashLoanEnabled) {
+            revert FlashLoanNotEnabled();
+        }
+
+        ensureNonZeroAddress(receiver);
+
+        // Check if the caller is authorized to execute flash loans
+        if (!comptroller.authorizedFlashLoan(initiator)) {
+            revert SenderNotAuthorized(initiator);
+        }
+
+        // Tracks the flashLoan amount before transferring amount to the receiver
+        flashLoanAmount += amount;
+
+        // Transfer the underlying asset to the receiver
+        doTransferOut(receiver, amount);
+
+        uint256 balanceBefore = getCashPrior();
+        (uint256 totalFee, uint256 protocolFee) = calculateFlashLoanFee(amount);
+        uint256 repayAmount = amount + totalFee;
+
+        // Call the execute operation on receiver contract
+        if (!IFlashLoanSimpleReceiver(receiver).executeOperation(underlying, amount, totalFee, msg.sender, param)) {
+            revert ExecuteFlashLoanFailed();
+        }
+
+        doTransferIn(receiver, repayAmount);
+        flashLoanAmount -= amount;
+
+        if ((getCashPrior() - balanceBefore) < repayAmount) revert InsufficientRepayment();
+
+        // Transfer protocol fee to protocol share reserve
+        doTransferOut(protocolShareReserve, protocolFee);
+
+        IProtocolShareReserve(protocolShareReserve).updateAssetsState(
+            address(comptroller),
+            underlying,
+            IProtocolShareReserve.IncomeType.FLASHLOAN
+        );
+
+        emit FlashLoanExecuted(receiver, underlying, amount);
+        return uint(Error.NO_ERROR);
+    }
+
+    /**
+     * @notice Enables or disables flash loan for the market
+     * @custom:access Only Governance
+     * @custom:event Emits ToggleFlashLoanEnabled event on success
+     */
+    function toggleFlashLoan() external returns (uint256) {
```

Comment:
I'd prefer to have a setter with an explicit `true` or `false` parameter. This way we can avoid race conditions between, e.g., a guardian transaction and a VIP transaction.